### PR TITLE
[SAP] Add vmware extension for migration

### PR DESCRIPTION
**Pull request by hemna**
This is a POC/WIP to add the extension to the cinder API
to allow calling os-migrate_volume_by_connector.  This will
still be an admin API endpoint as it will call migrate_volume
underneath it.

Notes:
  I tested this on my home devstack and was able to manually issue the call 
to migrate via connector.   It worked. :)

For using this manually you can do this:

```
connector = {"connection_capabilities": ["vmware_service_instance_uuid:......"]}
body = {"connector": connector}
client.volumes._action('os-migrate_volume_by_connector', volume, body)
```

### CODE REVIEW

File: cinder/api/contrib/vmware_migrate.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        The connector will contain the correct vcenter uuid
+        so that the scheduler can find the right cinder backend
+        to migrate the volume.
+        """
+        context = req.environ['cinder.context']
+        # Not found exception will be handled at the wsgi level
+        volume = self._get(context, id)
+        self.authorize(context, 'migrate_volume', target_obj=volume)
+        params = body['os-migrate_volume_by_connector']
+        connector = params.get('connector', {})
```
Comment (joker-at-work) [inline cinder/api/contrib/vmware_migrate.py:53]: I don't think this should allow empty values. It should raise a 400 if no connector was provided.

**Review by joker-at-work — COMMENTED** [review]
Looks good :+1:

File: cinder/api/contrib/vmware_migrate.py (hunk 1 above)
Comment (hemna) [reply cinder/api/contrib/vmware_migrate.py:53]: Fixed.

File: cinder/volume/api.py [hunk 2]
```diff
@@ -826,6 +827,43 @@ def migrate_volume_by_connector(self, ctxt, volume, connector):
...
+
+        # When the migration of an available volume starts, both the status
+        # and the migration status of the volume will be changed.
+        # If the admin sets lock_volume flag to True, the volume
+        # status is changed to 'maintenance', telling users
+        # that this volume is in maintenance mode, and no action is allowed
+        # on this volume, e.g. attach, detach, retype, migrate, etc.
+        if lock_volume:
+            updates['status'] = db.Case(
+                [(volume.model.status == 'available', 'maintenance')],
```
Comment (joker-at-work) [inline cinder/volume/api.py:855]: What about `reserved` here, don't we have to change to `maintenance` then, too?

File: cinder/volume/api.py [hunk 3]
```diff
@@ -826,6 +827,43 @@ def migrate_volume_by_connector(self, ctxt, volume, connector):
...
+                else_=volume.model.status)
+
+        result = volume.conditional_update(updates, expected, filters)
+
+        if not result:
+            msg = _('Volume %s status must be available or in-use, must not '
```
Comment (joker-at-work) [inline cinder/volume/api.py:861]: This is not true, right? It must be `available` or `reserved` as we set those in `expected`.

File: cinder/volume/api.py (hunk 2 above)
Comment (hemna) [reply cinder/volume/api.py]: done

File: cinder/volume/api.py (hunk 3 above)
Comment (hemna) [reply cinder/volume/api.py]: done

**Review by fwiesel — COMMENTED** [review]
I'm fine with it.

**Pull request merged**